---
title: '(December 2021) What I've been up to recently, and what I'm working on.'
date: 2021-12-25
permalink: /posts/2021/12/bp1/
tags:
  - Life
  - Ion channels
  - Electrophysiology
---

<style>
    pre {font-size: 12;}
    h3 {font-size: 1.35em;}
    h4 {font-size: 1.2em;}

    .inline-block {
        display: inline-block;
    }

    .container {
        max-width: 100vw; 
        margin: 0;
        horizontal-align: middle;
    }
</style>

It's been a while since I've updated my blog. I'd like to share a couple updates on what I've been doing, and what I hope to accomplish within the next month. I'll start in chronological order, making note of the approximate dates when I worked on things. 

Also, my laptop is acting up right now, so I can't add links (freezes while typing every $n$ characters...). I really need a new machine, don't I. If anyone is willing to hire me, or sponsor a new machine... of course not... right? It is Christmas, after all... 

# Part I: Exploration and computational experiments. 
By July 2021, I had completed my last, pre-defense committee meeting, and was fully committed on writing my thesis. However, there were several unturned stones that were bothering me, and in the absence of both intervention and calculated reasoning, I decided to spend my newfound time away from the lab in exploring these ideas. I was initially quite confident that I could complete what I wanted to do in a couple weeks. The subtext, of course, is that I didn't, and, since none of this work will be in my thesis nor any eventual publication, I wasted several months, and continue to face the emotional, physical, and financial consequences of this series of unfortunate decisions.

Regrets aside, what did I actually accomplish in this period (roughly July through to September)? 

## Molecular dynamics (MD)

### Introduction

By this, I mean that I used [OpenMM](https://openmm.org) and mostly-default [CHARMM-GUI](https://www.charmm-gui.org/) settings to perform all-atom simulations of WT and F431A HCN2 channels. The homology models were built using the [SwissModel](https://swissmodel.expasy.org/) server from HCN1 (PDB: 5U6O, 5U6P, 6UQF, and 6UQG). These pilot experiments were inspired by those of the isolated segment of Kv1.5 channels shown in [Labro et al. (2003)](https://pubmed.ncbi.nlm.nih.gov/13679372/). This study was interesting because it focused on the *PVP* hinge motif, the first proline of which is homologous to F431 in HCN2, and showed that this first proline was required for channel opening. Their MD simulations were fairly limited, at 10 ns in duration and involving only the four S6 segments. The authors suggested that the first proline in the *PVP* motif introduces a kink that allows the pore to dilate upon voltage-sensor activation. 

<br>
<figure>
  <center>
    <img 
      src="https://github.com/haganenoneko/haganenoneko.github.io/blob/master/_posts/figs/2021-12-25-figs/Labro-PVP sequence.jpg?raw=true"
      alt="Alignment of the S6 segment of Kv channels."
      title="Alignment of the S6 segment of Kv channels." width="50%" height="50%"
      style="horizontal-align:middle"
    />
  </center>
  <figcaption>
    <b> Fig. 1 - Alignment of the S6 segment of Kv channels. </b>
    <br>
    <p> Sequence alignment of Kv channels around the *PVP* motif, showing its conservation. Figure reproduced without permission from <a href="https://pubmed.ncbi.nlm.nih.gov/13679372/">Labro et al. (2003)</a>. </p>
  </figcaption>
</figure>
<br>

### Differences between HCN and Kv channel opening mechanisms
However, the mechanism of pore opening likely differ between HCN and Kv channels. There are several reasons for this, including:

#### 1. Different relative arrangements of the voltage-sensing domain (VSD) and pore domain (PD).
In HCN channels, VSDs interact with PDs of the *same* subunit, whereas in Kv channels, VSDs interact with PDs of *adjacent* subunits. These configurations are called 'non-domain swapped' and 'domain-swapped,' respectively. 

<br>
<figure>
  <center>
    <img 
      src="https://github.com/haganenoneko/haganenoneko.github.io/blob/master/_posts/figs/2021-12-25-figs/Barros-swapped.png?raw=true"
      alt="Comparison of domain-swapped and non-domain-swapped arrangements."
      title="Comparison of domain-swapped and non-domain-swapped arrangements." width="60%" height="60%"
      style="horizontal-align:middle"
    />
  </center>
  <figcaption>
    <b> Fig. 2 - Comparison of domain-swapped and non-domain-swapped arrangements. </b>
    <br>
    <p> Pores are coloured blue, S4-S5 linkers in magenta, and the voltage-sensing domain (VSD) in black. Orange marks positively charged S4 residues. Views are from the membrane plane (top) and from the cytosolic side (bottom). Figure reproduced without permission from <a href="https://pubmed.ncbi.nlm.nih.gov/32351384/">Barros et al. (2020)</a>. </p>
  </figcaption>
</figure>
<br>

#### 2. Differences in VSD-PD ('electromechanical') coupling.
Due to the different structural arrangements of the PD and VSD, the mechanism by which voltage-sensing is coupled to the pore is also different. However, the most striking evidence of this is that HCN channels activate upon hyperpolarization, rather than depolarization like most Kv channels, despite HCN channels being members of the latter. In fact, gating currents also show that the S4 segments also move similarly upon changes in membrane potential (e.g. [Mannikko et al. (2002)](https://pubmed.ncbi.nlm.nih.gov/12397358/), Ryu et al. (2012)). 

#### 3. Differences in opening energetics and kinetics.
The current model of HCN channel activation is that, in the depolarized state, the VSDs push the pore closed and are thus autoinhibitory for channel opening. This is supported by structures of HCN1 and HCN4  that show tight planar arrays formed by S4, S5, and S6 ([Lee and MacKinnon (2017)](https://pubmed.ncbi.nlm.nih.gov/28086084/), [Saponaro et al. (2021)](https://pubmed.ncbi.nlm.nih.gov/34166608/)). The next key idea comes from my Master's lab in an alanine/valine scan of 20-odd residues in the distal S6 of HCN2. In this study, Vincenzo ('Enzo') Macri, a PhD alum, found that the majority of mutants exhibited compromised opening ([Macri et al. (2009)](https://pubmed.ncbi.nlm.nih.gov/19269964/)). In contrast, similar studies in *Shaker* Kv channels showed the opposite, i.e. that most mutations in the distal pore resulted in greater opening at rest. A straightforward interpretation of these observations is that the energetics of pore opening differ between *Shaker* and HCN: the open state is intrinsically *less* favourable in *Shaker*, but *more* favourable in HCN chanels. 

Hyperpolarization-dependent activation in HCN channels can therefore be explained as follows. Though the open state of the pore is energetically favoured, it cannot open upon depolarization due to autoinhibition from the VSDs, which pack against and close the pore at rest (depolarization). However, when displaced by hyperpolarization, VSD-PD contact is weakened, allowing the pore to *relax* into an open state. In contrast, opening of Kv channels is *active* and requires S4s the 'pull' the S6 gate open (hence, 'electromechanical'), whereas closing is passive. 

Below are some additional, noteworthy observations regarding electromechanical coupling in HCN and Kv channels.

**a. Kv channels require an S4-S5 linker, but not HCN channels.**

Electromechanical coupling in Kv channels requires a short, intracellular linker between the S4 and S5 ('S4-S5 linker'), presumably to transduce the aforementioned force of 'pulling.' Conversely, deletion or truncation of the S4-S5 linker does not impair normal gating in HCN channels. In fact, this appears to be the case for a number of '[non-domain swapped](#different-relative-arrangements-of-the-voltage-sensing-domain-vsd-and-pore-domain-pd)' channels, including [hERG and rEAG (Lorinczi et al., 2015)](https://pubmed.ncbi.nlm.nih.gov/25818916/). 

**b. Weak VSD-PD coupling.**

VSD-PD coupling is *very* weak in HCN channels! For one, kinetic models have predicted allosteric coupling factors that are at least 100 times less than those of *Shaker* ([Ryu et al. (2012)](https://pubmed.ncbi.nlm.nih.gov/23071265/), [Flynn et al. (2018)](https://pubmed.ncbi.nlm.nih.gov/30076228/)). Secondly, many mutations of the S4 and S5 cause raise baseline conductance at positive voltages, or even result in constitutively open channels (e.g. [Cheng et al. (2007)](https://pubmed.ncbi.nlm.nih.gov/17578902/), [Wemhöner et al. (2012)](https://pubmed.ncbi.nlm.nih.gov/23048023/), [Ramentol et al. (2020)](https://pubmed.ncbi.nlm.nih.gov/32184399/)). This may seem surprising, given the result of mutations in the distal S6, but actually, this actually has a very intuitive explanation. 

Given the favourable energetics of pore opening, channels would be expected to open at positive voltages in the absence of either: (1) voltage-sensing or (2) VSD-PD interaction. Since VSD-PD coupling is weak and required for deactivation, but not activation, channel closing is easily compromised by relatively small perturbations. Perhaps the best demonstration of weak VSD-PD coupling is the ability to convert HCN channels into depolarization-activated channels with just a few mutations at the S4-S5 interface (Ramentol et al. (2020)). 

However, there still remains a lot of work to be done. Personally, I've always been more interested in the *kinetic* side of the story, as there are a number of unique kinetic features of HCN channels. In fact, I could write another blog post about these. For example, while HCN and hERG are non-domain swapped channels that don't require their S4-S5 linkers for normal voltage-gating, hERG not only exhibits depolarization-dependent activation, but also opens much faster than HCN2/HCN4 - less than a second and several seconds, respectively. 

So, VSD-PD coupling isn't the whole story! To explain these differences, we need to (re)-visit the S5 and S6. In recent years, we have uncovered much about how the S4 and S5 communicate, and how C-terminal regions of the channel interact with the VSD, but much remains unknown about the S6. How does it open? How does it close? I attempted to answer the latter in my Master's thesis (I'll add a link to it in six or so months when it's finally published). 

### Results of a *brief* set of MD experiments
As stated above, I performed a series of MD experiments using homology models of WT and F431A HCN2 channels. As mentioned above, F431 is homologous to the first proline in the *PVP* hinge motif. In HCN channels, this pheylalanine is located in the distal S6 and is located across all isoforms. My Master's thesis focused on understanding the biophysics of the F431A mutant using electrophysiology and mathematical modelling, but these methods naturally come short in providing a truly molecular interpretation. 

Like Labro et al. (2003), my simulations were 10 ns long in duration (following appropriate minimization and equilibration), but whereas Labro and colleagues only simulated the S5 segments, I performed all-atom simulations in which the channels were embedded and simulated within a POPC bilayer. 

For the sake of brevity (there's still *much* more to cover in this blog post...), I will only show summary metrics of these experiments. I show both the summary figures and relevant (partial) code used to generate the data, but not necessarily the visualizations. All code is written in Python 3.7 and assumes the follow packages are imported, and that the trajectory is centerd and aligned:

<pre style='color:#d1d1d1;background:#000000;'><span style='color:#e66170; font-weight:bold; '>import</span> numpy <span style='color:#e66170; font-weight:bold; '>as</span> np 
<span style='color:#e66170; font-weight:bold; '>import</span> pandas <span style='color:#e66170; font-weight:bold; '>as</span> pd 
<span style='color:#e66170; font-weight:bold; '>import</span> MDAnalysis <span style='color:#e66170; font-weight:bold; '>as</span> mda 

<span style='color:#e66170; font-weight:bold; '>def</span> RemoveOverallMovement<span style='color:#d2cd86; '>(</span>
    u<span style='color:#d2cd86; '>:</span> mda<span style='color:#d2cd86; '>.</span>Universe<span style='color:#d2cd86; '>,</span> u2<span style='color:#d2cd86; '>:</span> mda<span style='color:#d2cd86; '>.</span>Universe<span style='color:#d2cd86; '>,</span> out_path<span style='color:#d2cd86; '>:</span> <span style='color:#e66170; font-weight:bold; '>str</span>
<span style='color:#d2cd86; '>)</span><span style='color:#d2cd86; '>:</span>
    <span style='color:#9999a9; '>"""Align a trajectory to its first frame to remove overall movements</span>
<span style='color:#9999a9; '></span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;Args:</span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;u (mda.Universe): coordinates and trajectory </span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;u2 (mda.Universe): copy of `u`</span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;out_path (str): filename for saved alignment (dcd file)</span>
<span style='color:#9999a9; '></span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;Returns:</span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;mda.Universe: aligned mda Universe</span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;"""</span>
    u<span style='color:#d2cd86; '>.</span>trajectory<span style='color:#d2cd86; '>[</span><span style='color:#00dddd; '>-</span><span style='color:#00a800; '>1</span><span style='color:#d2cd86; '>]</span>
    u2<span style='color:#d2cd86; '>.</span>trajectory<span style='color:#d2cd86; '>[</span><span style='color:#00a800; '>0</span><span style='color:#d2cd86; '>]</span> 
    
    align<span style='color:#d2cd86; '>.</span>AlignTraj<span style='color:#d2cd86; '>(</span>u<span style='color:#d2cd86; '>,</span> u2<span style='color:#d2cd86; '>,</span> select<span style='color:#d2cd86; '>=</span><span style='color:#00c4c4; '>'name CA'</span><span style='color:#d2cd86; '>,</span> filename<span style='color:#d2cd86; '>=</span>out_path<span style='color:#d2cd86; '>)</span><span style='color:#d2cd86; '>.</span>run<span style='color:#d2cd86; '>(</span><span style='color:#d2cd86; '>)</span>
    
    <span style='color:#e66170; font-weight:bold; '>return</span> u 
</pre>
<!--Created using ToHtml.com on 2021-12-25 14:42:50 UTC -->

#### **Root Mean Squared Deviation (RMSD)** 
RMSDs were computed for the entire trajectory using the backbone carbons of each residue, and then averaged across the four subunits. RMSD trajectories are often used to estimate how close to steady-state a system is: the higher the RMSD, the more 'dynamic' the system and the further it is from steady-state. In my simulations, RMSDs seem to stabilize, or be very close to stabilizing, by the end of 10 ns. However, the S4 segment of 'F431A_5u6o' seems to be substantially more dynamic than other models simulated (see legend of Figure 2 for an explanation of the naming). This could indicate that the F431A mutation, despite being located on the S6, 

<br>
<figure>
  <center>
    <img 
      src="https://github.com/haganenoneko/haganenoneko.github.io/blob/master/_posts/figs/2021-12-25-figs/MD-rmsd.png?raw=true"
      alt="RMSDs of WT and F431A HCN2 channels.."
      title="RMSDs of WT and F431A HCN2 channels." width="90%" height="90%"
      style="horizontal-align:middle"
    />
  </center>
  <figcaption>
    <b> Fig. 4 - RMSDs of WT and F431A HCN2 channels. </b>
    <br>
    <p> RMSD over the 10-ns trajectory is shown for the full protein ('Full') and individually for segments S4, S5, and S6. RMSD values were averaged across subunits. Colours correspond to different protein templates: WT HCN2 ('wt_5u6o' and 'wt_6uqg'), F431A HCN2 ('F431A_5u6o' and 'F431A_6uqg'). The suffices correspond to the RCSB IDs of models used as homology templates, where 5U6O corresponds to the closed HCN1 channel (Lee and MacKinnon, 2017) and 6UQG to constitutively open HCN1 mutant (<a href="https://pubmed.ncbi.nlm.nih.gov/31787376/">Lee and MacKinnon, 2019</a>). </p>
  </figcaption>
</figure>
<br>

<pre style='color:#d1d1d1;background:#000000;'><span style='color:#e66170; font-weight:bold; '>def</span> do_RMSD<span style='color:#d2cd86; '>(</span>u<span style='color:#d2cd86; '>:</span> mda<span style='color:#d2cd86; '>.</span>Universe<span style='color:#d2cd86; '>,</span> selections<span style='color:#d2cd86; '>:</span> <span style='color:#e66170; font-weight:bold; '>list</span><span style='color:#d2cd86; '>,</span> sel_names<span style='color:#d2cd86; '>=</span><span style='color:#d2cd86; '>[</span><span style='color:#d2cd86; '>]</span><span style='color:#d2cd86; '>,</span> weights<span style='color:#d2cd86; '>=</span>None<span style='color:#d2cd86; '>)</span><span style='color:#d2cd86; '>:</span>
    <span style='color:#9999a9; '>"""perform RMSD on each frame of u.trajectory</span>
<span style='color:#9999a9; '></span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;Args:</span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;u (mda.Universe): coordinates and trajectory</span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;selections (list): list of strings conforming to MDAnalysis' atom selection language, e.g. 'resid 1-50' </span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;sel_names (list): list of strings representing names of groups in `selections` used as column headers for output dataframe. If empty, uses default numbers.</span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;weights (None or str, optional): weighting for RMSD calculation, of same length as number of particles in corresponding AtomGroup. Use 'mass' for weights of particle weights. Defaults to None.</span>
<span style='color:#9999a9; '></span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;Returns:</span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;pd.DataFrame: RMSD values </span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;"""</span>
    <span style='color:#9999a9; '># universe is automatically translated/centered </span>
    R <span style='color:#d2cd86; '>=</span> rms<span style='color:#d2cd86; '>.</span>RMSD<span style='color:#d2cd86; '>(</span>
        u<span style='color:#d2cd86; '>,</span> u<span style='color:#d2cd86; '>,</span> select<span style='color:#d2cd86; '>=</span><span style='color:#00c4c4; '>'backbone'</span><span style='color:#d2cd86; '>,</span> weights<span style='color:#d2cd86; '>=</span>weights<span style='color:#d2cd86; '>,</span>
        groupselections<span style='color:#d2cd86; '>=</span>selections<span style='color:#d2cd86; '>,</span> ref_frame<span style='color:#d2cd86; '>=</span><span style='color:#00a800; '>0</span>
    <span style='color:#d2cd86; '>)</span>
    
    R<span style='color:#d2cd86; '>.</span>run<span style='color:#d2cd86; '>(</span><span style='color:#d2cd86; '>)</span>
    
    <span style='color:#9999a9; '># column names for output dataframe</span>
    cols <span style='color:#d2cd86; '>=</span> <span style='color:#d2cd86; '>[</span><span style='color:#00c4c4; '>'Frame'</span><span style='color:#d2cd86; '>,</span> <span style='color:#00c4c4; '>'Time (ns)'</span><span style='color:#d2cd86; '>]</span>
    <span style='color:#e66170; font-weight:bold; '>if</span> <span style='color:#e66170; font-weight:bold; '>len</span><span style='color:#d2cd86; '>(</span>sel_names<span style='color:#d2cd86; '>)</span> <span style='color:#00dddd; '>&lt;</span> <span style='color:#00a800; '>1</span><span style='color:#d2cd86; '>:</span>
        cols<span style='color:#d2cd86; '>.</span>extend<span style='color:#d2cd86; '>(</span><span style='color:#e66170; font-weight:bold; '>list</span><span style='color:#d2cd86; '>(</span><span style='color:#e66170; font-weight:bold; '>range</span><span style='color:#d2cd86; '>(</span><span style='color:#e66170; font-weight:bold; '>len</span><span style='color:#d2cd86; '>(</span>selections<span style='color:#d2cd86; '>)</span><span style='color:#d2cd86; '>)</span><span style='color:#d2cd86; '>)</span><span style='color:#d2cd86; '>)</span>
    <span style='color:#e66170; font-weight:bold; '>else</span><span style='color:#d2cd86; '>:</span>
        cols<span style='color:#d2cd86; '>.</span>extend<span style='color:#d2cd86; '>(</span>sel_names<span style='color:#d2cd86; '>)</span>
    
    <span style='color:#e66170; font-weight:bold; '>return</span> pd<span style='color:#d2cd86; '>.</span>DataFrame<span style='color:#d2cd86; '>(</span>R<span style='color:#d2cd86; '>.</span>rmsd<span style='color:#d2cd86; '>,</span> columns<span style='color:#d2cd86; '>=</span>cols<span style='color:#d2cd86; '>)</span>
</pre>
<!--Created using ToHtml.com on 2021-12-25 14:43:55 UTC -->

#### **Root Mean Squared Fluctuation**
The RMSF is very similar to the RMSD if you imagine averaging the RMSD for each timepoint. The RMSF measures the movement of each residue throughout the trajectory. Comparing the same simulations as above, we see that the RMSFs are very similar, although 'F431A_5u6o' seems to be more dynamic between S4-S5 and just distal to the S6. 

<br>
<figure>
  <center>
    <img 
      src="https://github.com/haganenoneko/haganenoneko.github.io/blob/master/_posts/figs/2021-12-25-figs/MD-rmsf.png?raw=true"
      alt="RMSFs of WT and F431A HCN2 channels.."
      title="RMSFs of WT and F431A HCN2 channels." width="90%" height="90%"
      style="horizontal-align:middle"
    />
  </center>
  <figcaption>
    <b> Fig. 5 - RMSFs of WT and F431A HCN2 channels. </b>
    <br>
    <p> RMSFs over the 10-ns trajectory is shown for WT HCN2 ('wt_5u6o' and 'wt_6uqg') and F431A HCN2 ('F431A_5u6o' and 'F431A_6uqg'). See the legend of Figure 4 for an explanation of suffices. </p>
  </figcaption>
</figure>
<br>

<pre style='color:#d1d1d1;background:#000000;'>
<span style='color:#e66170; font-weight:bold; '>def</span> do_RMSF<span style='color:#d2cd86; '>(</span>
    u<span style='color:#d2cd86; '>:</span> mda<span style='color:#d2cd86; '>.</span>Universe<span style='color:#d2cd86; '>,</span> out_path<span style='color:#d2cd86; '>:</span> <span style='color:#e66170; font-weight:bold; '>str</span><span style='color:#d2cd86; '>,</span> 
    save_avg_pdb<span style='color:#d2cd86; '>=</span>False<span style='color:#d2cd86; '>,</span> save_avg_traj<span style='color:#d2cd86; '>=</span>False<span style='color:#d2cd86; '>,</span> 
    show<span style='color:#d2cd86; '>=</span>True<span style='color:#d2cd86; '>,</span> out<span style='color:#d2cd86; '>=</span>False
<span style='color:#d2cd86; '>)</span><span style='color:#d2cd86; '>:</span>
    <span style='color:#9999a9; '>"""Compute RMSF over trajectory relative to average structure</span>
<span style='color:#9999a9; '></span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;Args:</span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;u (mda.Universe): coordinates and trajectory loaded with MDAnalysis</span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;out_path (str): folder to hold output files </span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;save_avg_pdb (bool, optional): whether to save average pdb. Defaults to False.</span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;save_avg_traj (bool, optional): whether to save average-aligned trajectory. Defaults to False.</span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;show (bool, optional): whether to plot RMSF. Defaults to True.</span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;out (bool, optional): whether to write PDB coloured by RMSF (B factors). Defaults to False.</span>
<span style='color:#9999a9; '></span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;Returns:</span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;np.ndarray: 1D array of RMSF values for each frame </span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;"""</span>
    <span style='color:#9999a9; '># create average structure</span>
    avg <span style='color:#d2cd86; '>=</span> align<span style='color:#d2cd86; '>.</span>AverageStructure<span style='color:#d2cd86; '>(</span>
        u<span style='color:#d2cd86; '>,</span> u<span style='color:#d2cd86; '>,</span> select<span style='color:#d2cd86; '>=</span><span style='color:#00c4c4; '>'protein and name CA'</span><span style='color:#d2cd86; '>,</span> ref_frame<span style='color:#d2cd86; '>=</span><span style='color:#00a800; '>0</span>
    <span style='color:#d2cd86; '>)</span><span style='color:#d2cd86; '>.</span>run<span style='color:#d2cd86; '>(</span><span style='color:#d2cd86; '>)</span>
    
    <span style='color:#e66170; font-weight:bold; '>if</span> save_avg_pdb<span style='color:#d2cd86; '>:</span>
        avg<span style='color:#d2cd86; '>.</span>atoms<span style='color:#d2cd86; '>.</span>write<span style='color:#d2cd86; '>(</span>out_path <span style='color:#00dddd; '>+</span> <span style='color:#00c4c4; '>"_avg.pdb"</span><span style='color:#d2cd86; '>)</span>
        <span style='color:#e66170; font-weight:bold; '>print</span><span style='color:#d2cd86; '>(</span><span style='color:#00c4c4; '>"Saved averaged PDB to %s_avg.pdb"</span> <span style='color:#00dddd; '>%</span> out_path<span style='color:#d2cd86; '>)</span>
    
    <span style='color:#9999a9; '># align backbone of average structure and trajectory  </span>
    <span style='color:#e66170; font-weight:bold; '>if</span> save_avg_traj<span style='color:#d2cd86; '>:</span>
        align<span style='color:#d2cd86; '>.</span>AlignTraj<span style='color:#d2cd86; '>(</span>
            u<span style='color:#d2cd86; '>,</span> avg<span style='color:#d2cd86; '>.</span>universe<span style='color:#d2cd86; '>,</span> select<span style='color:#d2cd86; '>=</span><span style='color:#00c4c4; '>'protein and name CA'</span><span style='color:#d2cd86; '>,</span>
            filename <span style='color:#d2cd86; '>=</span> out_path <span style='color:#00dddd; '>+</span> <span style='color:#00c4c4; '>"_avg_aligned.dcd"</span>
        <span style='color:#d2cd86; '>)</span><span style='color:#d2cd86; '>.</span>run<span style='color:#d2cd86; '>(</span><span style='color:#d2cd86; '>)</span>
        <span style='color:#e66170; font-weight:bold; '>print</span><span style='color:#d2cd86; '>(</span><span style='color:#00c4c4; '>"Saved average-aligned trajectory to %s_avg_aligned.dcd"</span> <span style='color:#00dddd; '>%</span> out_path<span style='color:#d2cd86; '>)</span>
    <span style='color:#e66170; font-weight:bold; '>else</span><span style='color:#d2cd86; '>:</span>
        align<span style='color:#d2cd86; '>.</span>AlignTraj<span style='color:#d2cd86; '>(</span>u<span style='color:#d2cd86; '>,</span> avg<span style='color:#d2cd86; '>.</span>universe<span style='color:#d2cd86; '>,</span> select<span style='color:#d2cd86; '>=</span><span style='color:#00c4c4; '>'protein and name CA'</span><span style='color:#d2cd86; '>)</span><span style='color:#d2cd86; '>.</span>run<span style='color:#d2cd86; '>(</span><span style='color:#d2cd86; '>)</span>
    
    <span style='color:#9999a9; '># compute rmsf </span>
    c_alphas <span style='color:#d2cd86; '>=</span> u<span style='color:#d2cd86; '>.</span>select_atoms<span style='color:#d2cd86; '>(</span><span style='color:#00c4c4; '>'protein and name CA'</span><span style='color:#d2cd86; '>)</span>
    R <span style='color:#d2cd86; '>=</span> rms<span style='color:#d2cd86; '>.</span>RMSF<span style='color:#d2cd86; '>(</span>c_alphas<span style='color:#d2cd86; '>)</span><span style='color:#d2cd86; '>.</span>run<span style='color:#d2cd86; '>(</span><span style='color:#d2cd86; '>)</span>
    
    <span style='color:#e66170; font-weight:bold; '>if</span> show<span style='color:#d2cd86; '>:</span>
        plt<span style='color:#d2cd86; '>.</span>plot<span style='color:#d2cd86; '>(</span>c_alphas<span style='color:#d2cd86; '>.</span>resids<span style='color:#d2cd86; '>,</span> R<span style='color:#d2cd86; '>.</span>rmsf<span style='color:#d2cd86; '>)</span>
        plt<span style='color:#d2cd86; '>.</span>xlabel<span style='color:#d2cd86; '>(</span><span style='color:#00c4c4; '>'Residue number'</span><span style='color:#d2cd86; '>)</span>
        plt<span style='color:#d2cd86; '>.</span>ylabel<span style='color:#d2cd86; '>(</span><span style='color:#00c4c4; '>'RMSF ($</span><span style='color:#008080; '>\A</span><span style='color:#00c4c4; '>A$)'</span><span style='color:#d2cd86; '>)</span>
        plt<span style='color:#d2cd86; '>.</span>show<span style='color:#d2cd86; '>(</span><span style='color:#d2cd86; '>)</span>
    
    <span style='color:#9999a9; '>## add empty attribute for all atoms</span>
    <span style='color:#e66170; font-weight:bold; '>if</span> out<span style='color:#d2cd86; '>:</span>
        np<span style='color:#d2cd86; '>.</span>save<span style='color:#d2cd86; '>(</span>out_path <span style='color:#00dddd; '>+</span> <span style='color:#00c4c4; '>"rmsf.npy"</span><span style='color:#d2cd86; '>,</span> R<span style='color:#d2cd86; '>.</span>rmsf<span style='color:#d2cd86; '>)</span>
        <span style='color:#e66170; font-weight:bold; '>print</span><span style='color:#d2cd86; '>(</span><span style='color:#00c4c4; '>"RMSF saved to %s_rmsf.npy"</span> <span style='color:#00dddd; '>%</span> out_path<span style='color:#d2cd86; '>)</span>
        
        u<span style='color:#d2cd86; '>.</span>add_TopologyAttr<span style='color:#d2cd86; '>(</span><span style='color:#00c4c4; '>'tempfactors'</span><span style='color:#d2cd86; '>)</span> 
        protein <span style='color:#d2cd86; '>=</span> u<span style='color:#d2cd86; '>.</span>select_atoms<span style='color:#d2cd86; '>(</span><span style='color:#00c4c4; '>'protein'</span><span style='color:#d2cd86; '>)</span> <span style='color:#9999a9; '># select protein atoms</span>
        <span style='color:#e66170; font-weight:bold; '>for</span> residue<span style='color:#d2cd86; '>,</span> r_value <span style='color:#e66170; font-weight:bold; '>in</span> <span style='color:#e66170; font-weight:bold; '>zip</span><span style='color:#d2cd86; '>(</span>protein<span style='color:#d2cd86; '>.</span>residues<span style='color:#d2cd86; '>,</span> R<span style='color:#d2cd86; '>.</span>rmsf<span style='color:#d2cd86; '>)</span><span style='color:#d2cd86; '>:</span>
            residue<span style='color:#d2cd86; '>.</span>atoms<span style='color:#d2cd86; '>.</span>tempfactors <span style='color:#d2cd86; '>=</span> r_value
        u<span style='color:#d2cd86; '>.</span>atoms<span style='color:#d2cd86; '>.</span>write<span style='color:#d2cd86; '>(</span>out_path <span style='color:#00dddd; '>+</span> <span style='color:#00c4c4; '>'_rmsf.pdb'</span><span style='color:#d2cd86; '>)</span>
        <span style='color:#e66170; font-weight:bold; '>print</span><span style='color:#d2cd86; '>(</span><span style='color:#00c4c4; '>"RMSF-coloured PDB saved to %s_rmsf.pdb"</span> <span style='color:#00dddd; '>%</span> out_path<span style='color:#d2cd86; '>)</span>
    
    <span style='color:#e66170; font-weight:bold; '>return</span> R<span style='color:#d2cd86; '>.</span>rmsf 
</pre>
<!--Created using ToHtml.com on 2021-12-25 14:45:25 UTC -->

#### Radius of gyration (Rg)
Radius of gyration (Rg) is a familiar concept to physicists, but perhaps not for most (traditional) biologists. Wikipedia describes it as follows:

<br>
<blockquote cite="https://en.wikipedia.org/wiki/Radius_of_gyration" style="background-color:#3d3d3d; color:#ffffff"> the axis of rotation is defined as the radial distance to a point which would have a moment of inertia the same as the body's actual distribution of mass, if the total mass of the body were concentrated there </blockquote>
<br>

<br>
<figure>
  <center>
    <img 
      src="https://github.com/haganenoneko/haganenoneko.github.io/blob/master/_posts/figs/2021-12-25-figs/MD-rgyr.png?raw=true"
      alt="Rg of WT and F431A HCN2 channels.."
      title="Rg of WT and F431A HCN2 channels." width="90%" height="90%"
      style="horizontal-align:middle"
    />
  </center>
  <figcaption>
    <b> Fig. 6 - Radii of gyration of WT and F431A HCN2 channels. </b>
    <br>
    <p> Radii of gyration over the 10-ns trajectory is shown for WT HCN2 ('wt_5u6o' and 'wt_6uqg') and F431A HCN2 ('F431A_5u6o' and 'F431A_6uqg'). See the legend of Figure 4 for an explanation of suffices. Radii of gyration were averaged over each subunit and normalized by subtracting the mean of each time series. </p>
  </figcaption>
</figure>
<br>

<pre style='color:#d1d1d1;background:#000000;'><span style='color:#e66170; font-weight:bold; '>def</span> do_Rg<span style='color:#d2cd86; '>(</span>u<span style='color:#d2cd86; '>:</span> mda<span style='color:#d2cd86; '>.</span>Universe<span style='color:#d2cd86; '>)</span><span style='color:#d2cd86; '>:</span>
    <span style='color:#9999a9; '>"""Compute radius of gyration of backbone over trajectory</span>
<span style='color:#9999a9; '></span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;Args:</span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;u (mda.Universe): coordinates and trajectory</span>
<span style='color:#9999a9; '></span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;Returns:</span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;&#xa0;pd.DataFrame: radius of gyration with column of frames </span>
<span style='color:#9999a9; '>&#xa0;&#xa0;&#xa0;&#xa0;"""</span>
    prot <span style='color:#d2cd86; '>=</span> u<span style='color:#d2cd86; '>.</span>select_atoms<span style='color:#d2cd86; '>(</span><span style='color:#00c4c4; '>'protein and backbone'</span><span style='color:#d2cd86; '>)</span>
    
    frames <span style='color:#d2cd86; '>=</span> <span style='color:#d2cd86; '>[</span><span style='color:#d2cd86; '>]</span> 
    Rg <span style='color:#d2cd86; '>=</span> <span style='color:#d2cd86; '>[</span><span style='color:#d2cd86; '>]</span> 
    <span style='color:#e66170; font-weight:bold; '>for</span> ts <span style='color:#e66170; font-weight:bold; '>in</span> u<span style='color:#d2cd86; '>.</span>trajectory<span style='color:#d2cd86; '>:</span>
        Rg<span style='color:#d2cd86; '>.</span>append<span style='color:#d2cd86; '>(</span>prot<span style='color:#d2cd86; '>.</span>radius_of_gyration<span style='color:#d2cd86; '>(</span><span style='color:#d2cd86; '>)</span><span style='color:#d2cd86; '>)</span>
        frames<span style='color:#d2cd86; '>.</span>append<span style='color:#d2cd86; '>(</span>ts<span style='color:#d2cd86; '>.</span>frame<span style='color:#d2cd86; '>)</span>
        
    <span style='color:#e66170; font-weight:bold; '>return</span> pd<span style='color:#d2cd86; '>.</span>DataFrame<span style='color:#d2cd86; '>(</span><span style='color:#d2cd86; '>[</span>frames<span style='color:#d2cd86; '>,</span> Rg<span style='color:#d2cd86; '>]</span><span style='color:#d2cd86; '>,</span> columns<span style='color:#d2cd86; '>=</span><span style='color:#d2cd86; '>[</span><span style='color:#00c4c4; '>"Frame"</span><span style='color:#d2cd86; '>,</span> <span style='color:#00c4c4; '>"Rg"</span><span style='color:#d2cd86; '>]</span><span style='color:#d2cd86; '>)</span>
</pre>
<!--Created using ToHtml.com on 2021-12-25 15:13:04 UTC -->

#### Pore diameter analysis with [HOLE](http://www.holeprogram.org/)
The last of the "MD-specific" analyses I did was use HOLE to measure pore widths for each simulation. HOLE is an interesting program with a lot of history in the ion channel field. I don't really know how it works, but HOLE is used extensively to compute the width of ion channel pores, which I believe are located in structures using properties such as residue size and hydrophobicity. Some more recent tools have been developed, a bunch of which are available on web servers, e.g. [MOLEonline](https://mole.upol.cz/). I don't find these tools particularly novel or innovative, but I admittedly also don't understand them very well. Another software of this variety is the [Channel Annotation Package (CHAP)](https://www.channotation.org/) which was developed by Klesse and colleagues at Oxford ([Klesse et al. (2019)](https://doi.org/10.1016/j.jmb.2019.06.003)), although I first heard of it from [Rao et al. (2019)](https://www.pnas.org/content/116/28/13989). I like CHAP, since it's based on molecular dynamics simulations, and even tried it on my own several years ago, confirming the authors' results on HCN [Rao et al. (2019)](https://www.pnas.org/content/116/28/13989). 

<br>
<figure>
  <center>
    <img 
      src="https://github.com/haganenoneko/haganenoneko.github.io/blob/master/_posts/figs/2021-12-25-figs/MD-last hole profiles.png?raw=true"
      alt="Pore profiles of WT and F431A HCN2 channels.."
      title="Pore profiles of WT and F431A HCN2 channels." width="90%" height="90%"
      style="horizontal-align:middle"
    />
  </center>
  <figcaption>
    <b> Fig. 7 - Pore profiles of WT and F431A HCN2 channels. </b>
    <br>
    <p> Pore profiles over the 10-ns trajectory is shown for WT HCN2 ('wt_5u6o' and 'wt_6uqg') and F431A HCN2 ('F431A_5u6o' and 'F431A_6uqg'). See the legend of Figure 4 for an explanation of suffices. The pore profiles are roughly identical, suggesting that the mutation doesn't strongly affect pore radius. Pore widths are computed using the last frame of each trajectory. </p>
  </figcaption>
</figure>
<br>

### Network analyses with Cytoscape
After the initial MD analyses seemed to not give a clear indication of how F431A affected the dynamics of HCN2 channels, I turned to network-based analyses. An intuitive representation of a protein is as a graph with residues as nodes and bonds or interactions as edges. I used [Cytoscape](https://cytoscape.org/) and [RINalyzer](https://rinalyzer.de/) to build and analyze these graphs (called 'Residue Interaction Networks,' or RINs for short). The motivation was fairly simple: there may be effects of the mutation that are difficult to 'see' from movement-based metrics (e.g. RMSD, RMSF, etc.) alone. Indeed, some groups have combined MD and network analysis to look at allosteric signaling within proteins (e.g. [Fernandez-Marino et al. (2018)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6170002/) and [Kang et al. (2020)](https://www.science.org/doi/10.1126/sciadv.abd6798). 

This is totally irrelevant, but Po Wei Kang, the first author of the second paper, spoke at a meeting I attended in 2019. At the time, I was very much insulated from other ion channel researchers (still am, I guess). His presentation then was like a conversation you have during the day to which you think of a good reply hours or days after the conversation. Anyway, what I'm trying to say is that his work is superb. Somewhat reminds me of giants like Francisco Bezanilla or Clay Armstrong, but what do I know. 

Continuing on from where I left off - the application of network analysis on RINs. Network science is a very interesting developed field, and a lot of the results are intuitively integrated with fields like probability, electrical engineering, and dynamical systems. However, I only looked at a small subset of network science - network centrality. Centrality metrics all try to capture how 'important' a given node or edge is within a network. There are many ways to calculate centrality, including the number of neighbours a node has, the number of shortest paths that traverse a node, and similar definitions for edges. I started by computing different centrality metrics for each model (e.g. 'WT_5u6o', 'F431A_5u6o', etc.) to see whether there were differences between WT and F431A. Another aside - I know that this sounds like confirmation bias. It is, sort of. But I think about it as exploratory analysis, in the sense that I don't know much about my system (network properties of the protein) and that I have a bunch of drill bits (centrality metrics) that fit slightly differently (address different aspects of the network). So, this is more about tool selection. 

<br>
<figure class="inline-block">
  <center>
    <img 
      src="https://github.com/haganenoneko/haganenoneko.github.io/blob/master/_posts/figs/2021-12-25-figs/NetCent-5u6o_hcn2_Individual.png?raw=true"
      alt="Centrality metrics for WT_5u6o"
      title="Centrality metrics for WT_5u6o." width="100%" height=auto
    />
  </center>
</figure>
<br>


<div class="container">
    <figure>
        <img 
        src="https://github.com/haganenoneko/haganenoneko.github.io/blob/master/_posts/figs/2021-12-25-figs/NetCent-5u6o_hcn2_Grouped.png?raw=true"
        alt="Centrality metrics for WT_5u6o"
        title="Centrality metrics for WT_5u6o." width="80%" height=100%
        />
    </figure>
    <figure >
        <img 
        src="https://github.com/haganenoneko/haganenoneko.github.io/blob/master/_posts/figs/2021-12-25-figs/NetCent-6uqf_hcn2_Grouped.png?raw=true"
        alt="Centrality metrics for WT_5u6o"
        title="Centrality metrics for WT_5u6o." width="80%" height=auto
        />
    </figure>
    <figure >
        <img 
        src="https://github.com/haganenoneko/haganenoneko.github.io/blob/master/_posts/figs/2021-12-25-figs/NetCent-6uqg_hcn2_Grouped.png?raw=true"
        alt="Centrality metrics for WT_5u6o"
        title="Centrality metrics for WT_5u6o." width="80%" height=auto
        />
    </figure>
</div>

# Bayesian inference for an ion channel model

# References
Barros F, de la Peña P, Domínguez P, Sierra LM, Pardo LA. The EAG Voltage-Dependent K+ Channel Subfamily: Similarities and Differences in Structural Organization and Gating. Front Pharmacol. 2020;11:411. Published 2020 Apr 15. doi:10.3389/fphar.2020.00411

Cheng L, Kinard K, Rajamani R, Sanguinetti MC. Molecular mapping of the binding site for a blocker of hyperpolarization-activated, cyclic nucleotide-modulated pacemaker channels. J Pharmacol Exp Ther. 2007;322(3):931-939. doi:10.1124/jpet.107.121467

Fernández-Mariño AI, Harpole TJ, Oelstrom K, Delemotte L, Chanda B. Gating interaction maps reveal a noncanonical electromechanical coupling mode in the Shaker K+ channel. Nat Struct Mol Biol. 2018;25(4):320-326. doi:10.1038/s41594-018-0047-3

Flynn GE, Zagotta WN. Insights into the molecular mechanism for hyperpolarization-dependent activation of HCN channels. Proc Natl Acad Sci U S A. 2018;115(34):E8086-E8095. doi:10.1073/pnas.1805596115

Kang PW, Westerlund AM, Shi J, et al. Calmodulin acts as a state-dependent switch to control a cardiac potassium channel opening. Sci Adv. 2020;6(50):eabd6798. Published 2020 Dec 11. doi:10.1126/sciadv.abd6798

Klesse G, Rao S, Sansom MSP, Tucker SJ. CHAP: A Versatile Tool for the Structural and Functional Annotation of Ion Channel Pores. J Mol Biol. 2019;431(17):3353-3365. doi:10.1016/j.jmb.2019.06.003

Labro AJ, Raes AL, Bellens I, Ottschytsch N, Snyders DJ. Gating of shaker-type channels requires the flexibility of S6 caused by prolines. J Biol Chem. 2003;278(50):50724-50731. doi:10.1074/jbc.M306097200

Lee CH, MacKinnon R. Structures of the Human HCN1 Hyperpolarization-Activated Channel. Cell. 2017;168(1-2):111-120.e11. doi:10.1016/j.cell.2016.12.023

Macri V, Nazzari H, McDonald E, Accili EA. Alanine scanning of the S6 segment reveals a unique and cAMP-sensitive association between the pore and voltage-dependent opening in HCN channels. J Biol Chem. 2009;284(23):15659-15667. doi:10.1074/jbc.M809164200

Männikkö R, Elinder F, Larsson HP. Voltage-sensing mechanism is conserved among ion channels gated by opposite voltages. Nature. 2002;419(6909):837-841. doi:10.1038/nature01038

Ramentol R, Perez ME, Larsson HP. Gating mechanism of hyperpolarization-activated HCN pacemaker channels. Nat Commun. 2020;11(1):1419. Published 2020 Mar 17. doi:10.1038/s41467-020-15233-9

Rao S, Klesse G, Lynch CI, Tucker SJ, Sansom MSP. Molecular Simulations of Hydrophobic Gating of Pentameric Ligand Gated Ion Channels: Insights into Water and Ions. J Phys Chem B. 2021;125(4):981-994. doi:10.1021/acs.jpcb.0c09285

Ryu S, Yellen G. Charge movement in gating-locked HCN channels reveals weak coupling of voltage sensors and gate. J Gen Physiol. 2012;140(5):469-479. doi:10.1085/jgp.201210850

Saponaro A, Bauer D, Giese MH, et al. Gating movements and ion permeation in HCN4 pacemaker channels. Mol Cell. 2021;81(14):2929-2943.e6. doi:10.1016/j.molcel.2021.05.033

Wemhöner K, Silbernagel N, Marzian S, et al. A leucine zipper motif essential for gating of hyperpolarization-activated channels. J Biol Chem. 2012;287(48):40150-40160. doi:10.1074/jbc.M112.378513
